[core][Kotlin] Add cpp part of async functions

lukmccall opened the pull request:
# Why

Part of https://github.com/expo/expo/pull/16977.
Adds a cpp part of the async function support. 

# How

My implementation is very similar to the TurboModule implementation  - https://github.com/facebook/react-native/blob/main/ReactCommon/react/nativemodule/core/platform/android/ReactCommon/JavaTurboModule.cpp.

- Stored pointer to the js and native callback invoker. The second one is unused right now, but we will need it in the future. 
- Allow modules to register async function on their CPP representation.
- Allow CPP code to call module function with promise. 

# Test Plan

- bare-expo and simple app:
```kotlin
// Cellular module
function("test") { s: String -> s }.runSynchronously()
asyncFunction("testAsync") { s: String -> s }
```

```js
// App.js
const module = global.ExpoModules.ExpoCellular;
console.log(module.test);
console.log(global.ExpoModules.ExpoCellular == global.ExpoModules.ExpoCellular);
console.log(module.test == module.test);
console.log(module.test("abc"));

console.log(module.testAsync("abc").then((resutl) => console.log(resutl) ))
```

lukmccall committed:
[core][Kotlin] Add cpp part of async functions

barthap reviewed (approved):
LGTM 👍 
Personally, I'd add some brief doc comments to all these callback/promise related stuff to make understanding the call flow easier, but that's a nit, I managed to understand it anyway 😉

lukmccall committed:
[sweet api][Kotlin] Dispatch async cpp calls through the native module queue (#17056)

# Why

a follow up to the https://github.com/expo/expo/pull/17020.

# How

- Created a single thread queue for dispatching the functions of the native module. 
- Dispatch CPP async call to the created queue

# Test Plan

- bare-expo ✅
